hugr-passes/src/force_order.rs

```diff
@@ -0,0 +1,307 @@
+/// All dataflow parents which are transitive children of `root`, including
+/// `root` itself, will have their dataflow regions ordered.
+///
+/// Dataflow regions are ordered by inserting order edges between their
+/// immediate children. A dataflow parent with `C` children will have at most
+/// `C-1` edges added. Any node than can be ordered will be.
+///
+/// Nodes are ordered according to the `rank` function. Nodes of lower rank will
+/// be ordered earlier in their parent. Note that if  `rank(n1) < rank(n2)` it
+/// is not guaranteed that `n1` will be ordered before `n2`. If `n1` dominates
+/// `n2` it cannot be ordered after `n1` without invalidating `hugr`.  Nodes of
+/// equal rank will be ordered arbitrarily, although that arbitrary order is
+/// deterministic.
+pub fn force_order(
+    hugr: &mut impl HugrMut,
+    root: Node,
+    rank: impl Fn(Node) -> i64,
+) -> Result<(), HugrError> {
+    force_order_by_key(hugr, root, rank)
+}
+
+/// As [force_order], but allows a generic [Ord] choice for the result of the
+/// `rank` function.
+pub fn force_order_by_key<K: Ord>(
+    hugr: &mut impl HugrMut,
+    root: Node,
+    rank: impl Fn(Node) -> K,
+) -> Result<(), HugrError> {
+    let dataflow_parents = DescendantsGraph::<Node>::try_new(hugr, root)?
+        .nodes()
+        .filter(|n| hugr.get_optype(*n).tag() <= OpTag::DataflowParent)
+        .collect_vec();
+    for dp in dataflow_parents {
+        let sg = SiblingGraph::<Node>::try_new(hugr, dp)?;
+        let petgraph = NodeFiltered::from_fn(sg.as_petgraph(), |x| x != dp);
+        let ordered_nodes = ForceOrder::new(&petgraph, &rank)
+            .iter(&petgraph)
+            .filter(|&x| hugr.get_optype(x).tag() <= OpTag::DataflowChild)
+            .collect_vec();
+
+        // ordered_nodes must always include the input and output nodes, so the
+        // slice below is in bounds.
+        for (&n1, &n2) in zip(&ordered_nodes[..], &ordered_nodes[1..]) {
+            let (n1_ot, n2_ot) = (hugr.get_optype(n1), hugr.get_optype(n2));
+            assert_eq!(
+                Some(EdgeKind::StateOrder),
+                n1_ot.other_port_kind(Direction::Outgoing)
+            );
+            assert_eq!(
+                Some(EdgeKind::StateOrder),
+                n2_ot.other_port_kind(Direction::Incoming)
+            );
+            if !hugr.output_neighbours(n1).contains(&n2) {
+                hugr.connect(
+                    n1,
+                    n1_ot.other_output_port().unwrap(),
+                    n2,
+                    n2_ot.other_input_port().unwrap(),
+                );
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// An adaption of [petgraph::visit::Topo]. We differ only in that we sort nodes
+/// by the rank function before adding them to the internal work stack. This
+/// ensures we visit lower ranked nodes before higher ranked nodes whenever the
+/// topology of the graph allows.
+#[derive(Clone)]
+struct ForceOrder<K: Ord, N: Ord, VM, F> {
```

Comment:
No need to repeat the bound here (it's only needed in the `impl`)
```suggestion
struct ForceOrder<K, N, VM, F> {
```